File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofModels/FeatureRequest.swift
import Foundation

/// Codable mirror of the `feature_requests` row in Main Supabase. Field names
/// match the RN type in `wagerproof-mobile/app/(drawer)/(tabs)/feature-requests.tsx`
/// byte-for-byte so the same JSON payload decodes cleanly here.
///
/// Status semantics (from RN + RLS policies):
/// - `pending` — submitted but not yet approved by an editor (hidden from
///   regular users; admins-only view).
/// - `approved` — visible in the "Community Voting" section, users can vote.
/// - `roadmap` — promoted from community to the developer roadmap. Voting UI
///   is hidden; only a "N votes" badge is shown. Substate carried by
///   `roadmap_status`.
public struct FeatureRequest: Codable, Identifiable, Sendable, Hashable {
    public enum Status: String, Codable, Sendable, Hashable {
        case pending
        case approved
        case roadmap
    }

    /// Roadmap substate. Only populated when `status == .roadmap`.
    public enum RoadmapStatus: String, Codable, Sendable, Hashable {
        case planned
        case inProgress = "in_progress"
        case completed
    }

    public let id: String
    public let title: String
    public let description: String
    public let submittedBy: String
    public let submitterDisplayName: String
    public let status: Status
    public let roadmapStatus: RoadmapStatus?
    public let upvotes: Int
    public let downvotes: Int
    public let createdAt: String
    public let updatedAt: String

    /// Net votes = upvotes - downvotes. Used everywhere the RN screen shows
    /// the green/red signed badge.
    public var netVotes: Int { upvotes - downvotes }

    public init(
        id: String,
        title: String,
        description: String,
        submittedBy: String,
        submitterDisplayName: String,
        status: Status,
        roadmapStatus: RoadmapStatus? = nil,
        upvotes: Int,
        downvotes: Int,
        createdAt: String,
        updatedAt: String
    ) {
        self.id = id
        self.title = title
        self.description = description
        self.submittedBy = submittedBy
        self.submitterDisplayName = submitterDisplayName
        self.status = status
        self.roadmapStatus = roadmapStatus
        self.upvotes = upvotes
        self.downvotes = downvotes
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case description
        case submittedBy = "submitted_by"
        case submitterDisplayName = "submitter_display_name"
        case status
        case roadmapStatus = "roadmap_status"
        case upvotes
        case downvotes
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

/// Per-user vote record. Mirrors the `feature_request_votes` table.
/// One row per (user, feature_request) pair — re-voting the same type
/// deletes the row, switching type updates it. See `FeatureRequestsStore.vote`
/// for the toggle/switch semantics that match RN.
public struct FeatureRequestVote: Codable, Sendable, Hashable {
    public enum VoteType: String, Codable, Sendable, Hashable {
        case upvote
        case downvote
    }

    public let featureRequestId: String
    public let voteType: VoteType

    public init(featureRequestId: String, voteType: VoteType) {
        self.featureRequestId = featureRequestId
        self.voteType = voteType
    }

    enum CodingKeys: String, CodingKey {
        case featureRequestId = "feature_request_id"
        case voteType = "vote_type"
    }
}

File: wagerproof-android-native/core/models/src/main/kotlin/com/wagerproof/core/models/FeatureRequest.kt
package com.wagerproof.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * `feature_requests` row in main Supabase.
 *
 * Status semantics (from RN + RLS policies):
 * - `pending` — submitted but not yet approved (admins-only view).
 * - `approved` — visible in "Community Voting", users can vote.
 * - `roadmap` — promoted to the developer roadmap; voting UI hidden,
 *   substate carried by `roadmap_status`.
 */
@Serializable
data class FeatureRequest(
    val id: String,
    val title: String,
    val description: String,
    @SerialName("submitted_by") val submittedBy: String,
    @SerialName("submitter_display_name") val submitterDisplayName: String,
    val status: Status,
    @SerialName("roadmap_status") val roadmapStatus: RoadmapStatus? = null,
    val upvotes: Int,
    val downvotes: Int,
    @SerialName("created_at") val createdAt: String,
    @SerialName("updated_at") val updatedAt: String,
) {
    /** Net votes = upvotes - downvotes (the signed green/red badge). */
    val netVotes: Int get() = upvotes - downvotes

    @Serializable
    enum class Status(val raw: String) {
        @SerialName("pending") PENDING("pending"),
        @SerialName("approved") APPROVED("approved"),
        @SerialName("roadmap") ROADMAP("roadmap"),
    }

    /** Roadmap substate. Only populated when `status == ROADMAP`. */
    @Serializable
    enum class RoadmapStatus(val raw: String) {
        @SerialName("planned") PLANNED("planned"),
        @SerialName("in_progress") IN_PROGRESS("in_progress"),
        @SerialName("completed") COMPLETED("completed"),
    }
}

/**
 * Per-user vote record (`feature_request_votes` table). One row per
 * (user, feature_request) pair — re-voting the same type deletes the row,
 * switching type updates it (service-layer semantics).
 */
@Serializable
data class FeatureRequestVote(
    @SerialName("feature_request_id") val featureRequestId: String,
    @SerialName("vote_type") val voteType: VoteType,
) {
    @Serializable
    enum class VoteType(val raw: String) {
        @SerialName("upvote") UPVOTE("upvote"),
        @SerialName("downvote") DOWNVOTE("downvote"),
    }
}
